File: OJ/gdgzoi/primary/contest/2493/2493_3_correct.cc
#include <algorithm>
#include <iostream>
#include <vector>

const int INF = 0x7fffffff;
const int MAX_N = 2e3 + 10;

struct e
{
    int u;
    int v;
    int w;
};

std::vector<e> edge;
int fa[MAX_N];

inline int find(int x)
{
    if (fa[x] == x)
        return x;
    return fa[x] = find(fa[x]);
}

inline void merge(int x, int y)
{
    int fx = find(x);
    int fy = find(y);

    fa[fx] = fy;
}

int main()
{
    std::ios::sync_with_stdio(false);
    std::cin.tie(0);

    int n, m;
    std::cin >> n >> m;

    for (int i = 0; i < m; i++)
    {
        int u, v, w;
        std::cin >> u >> v >> w;

        edge.push_back((e){u, v, w});
    }

    std::sort(edge.begin(), edge.end(), [](e a, e b) { return a.w < b.w; });

    int minn = INF;
    for (int i = 0; i < m; i++)
    {
        for (int j = 0; j < MAX_N; j++)
            fa[j] = j;

        for (int j = i; j < m; j++)
        {
            merge(edge[j].u, edge[j].v);

            if (find(1) == find(n))
            {
                minn = std::min(minn, edge[j].w - edge[i].w);
                break;
            }
        }
    }

    std::cout << minn;
    return 0;
}

File: OJ/gdgzoi/primary/contest/2493/2493_3_correct.py
n, m = map(int, input().split())

edge = list()
for i in range(0, n):
    u, v, w = map(int, input().split())

    tmp = {"u": u, "v": v, "w": w}
    edge.append(tmp)

edge.sort(key=lambda e: e["w"])

MAX_N = int(2e3 + 10)
fa = list()
for i in range(0, MAX_N):
    fa.append(i)


def find(x):
    if fa[x] == x:
        return x

    fa[x] = find(fa[x])
    return fa[x]


def merge(x, y):
    fx = find(x)
    fy = find(y)

    fa[fx] = fy


minn = 0x7fffffff
for i in range(0, m):
    for j in range(0, MAX_N):
        fa[j] = j

    for j in range(i, m):
        merge(edge[j]["u"], edge[j]["v"])

        if find(1) == find(n):
            minn = min(minn, edge[j]["w"] - edge[i]["w"])
            break

print(minn)
